### DISCUSSION TITLE
Visual effects (aka "Juice")

### DISCUSSION THREAD

**Discussion by ZenithStar** [ramokz/phantom-camera#162] [Ideas]
Would it be within the scope of this projects to include features such as:
- Screen Shake
- Post-processing Motion Blur (e.g. https://github.com/davisbrandon02/motion-blur-godot-4.0)
- Dynamic FoV proportional to camera velocity
- Wind particle

or should these be a separate addon? If added to the main project, they would obviously be optionally enabled.

Inspiration:
https://www.youtube.com/watch?v=0JVM7mbVh6U

**Comment by ramokz** [comment]
> Screen Shake

Yes! It's planned for the upcoming v0.7 (#126)

> Post-processing Motion Blur

Hadn't thought about it, but it's a neat idea. By the looks of it, there is an [ongoing proposal for adding it in natively](https://github.com/godotengine/godot-proposals/issues/2933). So if that comes to pass, then it feels like it would be both better and easier to utilize that as it would be less overhead. Though, there is also an argument to allow a slot to apply a custom shader for the camera per `PCam` instance.

> Dynamic FoV proportional to camera velocity

Also hadn't thought of this. Though, am not too sure how useful this one will be to directly support. Since you could technically just tween between different PCams with different FoVs when the velocity reaches a threshold. Alternatively, you could also dynamically interpolate the FoV with `pcam.set_camera_fov(value)`, which shouldn't be particularly tricky.

> Wind particles

"Wind particles" specifically, to me, feels like a very project and art style specific need. Think, similar to the other point about Motion Blur, it's something that could be part of a general particle camera effect slot for each `PCam`. That said, am not too familiar with Godot Particles, so that's more of an open question for what an appropriate solution would look and work like.

**Reply by ZenithStar** [reply]
Okay cool. A good motion blur implementation looks to be quite a bit more complicated than the shader I linked. Hopefully maybe someday someone will get it into the core engine.

As for the last two items, I suppose they should be their own package and doesn't necessary need to depend on Phantom Camera. There's a few other things that I would also like to add that certainly wouldn't be appropriate for this addon, such as a wind "woosh" SFX when the camera blasts past nearby collisions, and possibly also a doppler effect pitch shift.
